### DISCUSSION TITLE
Welcome to MSLab Discussions!

### DISCUSSION THREAD

**Discussion by jaromirk** [Microsoft/WSLab#515] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by LevT2** [comment]
Hi @jaromirk 

If I refactor LabConfigs in such a way:

```
$LabConfigCommon = @{
    DomainAdminName = 'LabAdmin'; AdminPassword = 'LS1setup!';
    DCEdition = '4';
    
    # TelemetryLevel='Full' ; TelemetryNickname='' ;
    AdditionalNetworksConfig = @();
    VMs = @();    
}

$LabConfig = $LabConfigCommon + @{
    Prefix     = 'WSLab17763.973-'; 
    SwitchName = 'LabSwitch';
    Internet   = $true ; 
}

$LabConfig.VMs += @{ VMName = 'Management' ; ParentVHD = 'Win2019_G2.vhdx' }
$LabConfig.VMs += @{ VMName = 'CA'         ; ParentVHD = 'Win2019Core_G2.vhdx' }
$LabConfig.VMs += @{ VMName = 'Grafana'    ; ParentVHD = 'Win2019Core_G2.vhdx' }
$LabConfig.VMs += @{ VMName = 'Collector'  ; ParentVHD = 'Win2019Core_G2.vhdx' }
```

- would you accept such a PR? 
There must be more observability for newcomers like me.

**Reply by machv** [reply]
Hello @LevT2, what would be the benefit of separating LabConfig to these two hash tables?

Idea of LabConfig stored as native PowerShell file instead of JSON/XML is to provide full flexibility in structuring the configuration.

So using your proposed way is completely valid syntax and I am interested in the benefits of this.

**Reply by LevT2** [reply]
Hello @machv
I understand that idea, trying to break the simplicity not so much.

Nonetheless, my own initial understanding of the Labs would immediately benefit from shorter lines and some cosmetic changes in the labconfigs. I'd split also the VMs configuration within the exist hashtable, if it spans more than 2 or 3 entries.

I prefer a little DRY just to reduce the clutter a bit., supposing there was no focused effort to unify LabConfigs.

**Reply by LevT2** [reply]
BTW, the VM size defaults were implicitly set to minimums as of several years ago, yet some LabConfigs have commented out variants with e.g.. NestedVirt enabled.
Perhaps there are more options nowadays, and the default VM sizing should also be made more flexible? I'd rather not to manually edit every single parameter just to adapt every VM to my $900 Chinese assembly freedom, having 256Gb RAM onboard and dozens of CPU cores.

**Reply by machv** [reply]
So, the main goal would be to set some VM-level defaults (sizing, TMP etc.)?

```powershell
$vmDefaults = @{
  MemoryStartupBytes = 4GB
  VMProcessorCount = 4
  ParentVHD = 'Win2022Core_G2.vhdx'
}

$LabConfig.VMs += ($vmDefaults + @{ VMName = 'ca' } )
$LabConfig.VMs += ($vmDefaults + @{ VMName = 'web' } )
```

I can include this in the sample `LabConfig.ps1` file, original idea was to provide config as simple as possible, and yes as you mentioned, most laptops aren't powerful enough to run multiple VMs so default sizing is low.

**Reply by LevT2** [reply]
I understand your conservatism. Please consider also adding the $labDefaults, to extract these defaults from the existing mess with locally relevant entries. The Prefix, Internet and LabSwitch must be provided locally in many labs and nothing more, you may even force their presence. (There are also several occasional instances in the exist sample codebase, such as SwitchNics or ManagementSubnetIDs) 

Sorry, I'm still looking at it as a newcomer. I also understand you hesitating to make the labconfigs consistent as they come from different authors and regressions are possible.

**Reply by LevT2** [reply]
```$LabConfig.VMs += ($vmDefaults + @{ VMName = 'ca' } )```

There is a big caveat here with duplicate entries, picky manual work just to get the lab working.  So, the whole codebase adoption would immediately benefit from splitting ang grouping config entries.

**Reply by LevT2** [reply]
@machv
```
$LabConfigCommon = @{
    DomainAdminName = 'LabAdmin'; AdminPassword = 'LS1setup!';
    DCEdition = '4';
    
    # TelemetryLevel='Full' ; TelemetryNickname='' ;
    AdditionalNetworksConfig = @();
    VMs = @();    
}

$LabConfig = $LabConfigCommon + @{ 
    Prefix = 'MSLab-'                           # should make obligatory? move to $LabConfigCommon?
    Internet = $true ;
}

#2 nodes for AzSHCI Cluster
1..2 | ForEach-Object {
    $LABConfig.VMs += @{
        VMName = "ArcVMs$_" ; ParentVHD = 'AzSHCI21H2_G2.vhdx' ; 
        
        Configuration = 'S2D' ;                 #region storage config
        HDDNumber = 4 ; HDDSize= 4TB ;
        
        MemoryStartupBytes= 14GB;               #region sizing
        VMProcessorCount="Max" ;
        
        NestedVirt=$true ; VirtualTPM=$true     #region hardware features

        <# DisableWCF etc. #>                   #region software features  
    }
}
```

Do you mind this? Yep, this is a halfway to proper templating, graspable to newcomers too.

**Comment by webmasterfcn** [comment]
why is this not working?

    
    $LabConfig=@{
        DomainAdminName="LabAdmin";                  # Used during 2_CreateParentDisks (no affect if changed after this step)
        AdminPassword="LS1setup!";                   # Used during 2_CreateParentDisks. If changed after, it will break the functionality of 3_Deploy.ps1
        Prefix = "MSLab-";                           # (Optional) All VMs and vSwitch are created with this prefix, so you can identify the lab. If not specified, Lab folder name is used
        SwitchName = "LabSwitch";                    # (Optional) Name of vSwitch
        SwitchNICs = "NIC1","NIC2";                  # (Optional) Adds these NICs to vSwitch (without connecting hostOS).
        SecureBoot=$true;                            # (Optional) Useful when testing unsigned builds (Useful for MS developers for daily builds)
        DCEdition="4";                               # 4 for DataCenter or 3 for DataCenterCore
        InstallSCVMM="No";                           # (Optional) Yes/Prereqs/SQL/ADK/No
        AdditionalNetworksInDC=$false;               # (Optional) If Additional networks should be added also to DC
        DomainNetbiosName="dc";                    # (Optional) If set, custom domain NetBios name will be used. if not specified, Default "corp" will be used
        DomainName="cics.intern";               # (Optional) If set, custom DomainName will be used. If not specified, Default "Corp.contoso.com" will be used
        DefaultOUName="Workshop";                    # (Optional) If set, custom OU for all machines and account will be used. If not specified, default "Workshop" is created
        AllowedVLANs="1-10";                         # (Optional) Sets the list of VLANs that can be used on Management vNICs. If not specified, default "1-10" is set.
        Internet=$false;                             # (Optional) If $true, it will add external vSwitch and configure NAT in DC to provide internet (Logic explained below)
        UseHostDnsAsForwarder=$false;                # (Optional) If $true, local DNS servers will be used as DNS forwarders in DC
        CustomDnsForwarders=@("8.8.8.8","1.1.1.1");  # (Optional) If configured, script will use those servers as DNS fordwarders in DC (Defaults to 8.8.8.8 and 1.1.1.1)
        PullServerDC=$true;                          # (Optional) If $false, then DSC Pull Server will not be configured on DC
        ServerISOFolder="";                          # (Optional) If configured, script will use ISO located in this folder for Windows Server hydration (if more ISOs are present, then out-grid view is called)
        ServerMSUsFolder="";                         # (Optional) If configured, script will inject all MSU files found into server OS
        EnableGuestServiceInterface=$false;          # (Optional) If True, then Guest Services integration component will be enabled on all VMs.
        DCVMProcessorCount=2;                        # (Optional) 2 is default. If specified more/less, processorcount will be modified.
        DHCPscope="10.0.0.0";                        # (Optional) 10.0.0.0 is configured if nothing is specified. Scope has to end with .0 (like 10.10.10.0). It's always /24       
        DCVMVersion="9.0";                           # (Optional) Latest is used if nothing is specified. Make sure you use values like "8.0","8.3","9.0"
        TelemetryLevel="";                           # (Optional) If configured, script will stop prompting you for telemetry. Values are "None","Basic","Full"
        TelemetryNickname="";                        # (Optional) If configured, telemetry will be sent with NickName to correlate data to specified NickName. So when leaderboards will be published, MSLab users will be able to see their own stats
        AutoStartAfterDeploy=$false;                 # (Optional) If $false, no VM will be started; if $true or 'All' all lab VMs will be started after Deploy script; if 'DeployedOnly' only newly created VMs will be started.
        InternetVLAN="";                             # (Optional) If set, it will apply VLAN on Interent adapter connected to DC
        ManagementSubnetIDs="";                      # (Optional) If set, it will add another dhcp-enable management networks.
        Linux=$false;                                # (Optional) If set to $true, required prerequisities for building Linux images with Packer will be configured.
        LinuxAdminName="linuxadmin";                 # (Optional) If set, local user account with that name will be created in Linux image. If not, DomainAdminName will be used as a local account.
        SshKeyPath="$($env:USERPROFILE)\.ssh\id_rsa" # (Optional) If set, specified SSH key will be used to build and access Linux images.
        AutoClosePSWindows=$false;                   # (Optional) If set, the PowerShell console windows will automatically close once the script has completed successfully. Best suited for use in automated deployments.
        AutoCleanUp=$false;                          # (Optional) If set, after creating initial parent disks, files that are no longer necessary will be cleaned up. Best suited for use in automated deployments.
        AdditionalNetworksConfig=@();                # Just empty array for config below
        VMs=@();                                     # Just empty array for config below
    }

**Reply by machv** [reply]
In order to change a AD domain you can try this labconfig

```powershell
$LabConfig=@{ 
  DomainAdminName='LabAdmin'; AdminPassword='LS1setup!';DCEdition='4'; 
  AdditionalNetworksConfig=@(); VMs=@();
  DomainNetbiosName="EVAL";
  DomainName="eval.kohfinkhatsgefunden.de"; 
}

# Windows Server 2022
1..4 | ForEach-Object {$VMNames="S2D"; $LABConfig.VMs += @{ VMName = "$VMNames$_" ; Configuration = 'S2D' ; ParentVHD = 'Win2022Core_G2.vhdx'; SSDNumber = 0; SSDSize=800GB ; HDDNumber = 12; HDDSize= 4TB ; MemoryStartupBytes= 512MB }}
```

**Reply by webmasterfcn** [reply]
This Lab is okay but no internet
![Screenshot 2022-10-19 192304](https://user-images.githubusercontent.com/54039769/196762008-116feb2f-c915-4888-b4a3-358e3dfae161.png)

**Reply by jaromirk** [reply]
Add internet=$true and just run deploy again. It will add internet to existing lab

**Reply by webmasterfcn** [reply]
$LabConfig=@{ 
  DomainAdminName='LabAdmin'; AdminPassword='LS1setup!';DCEdition='4'; 
  AdditionalNetworksConfig=@(); VMs=@();
  Internet=$true	
  DomainNetbiosName="EVAL";
  DomainName="eval.kohfinkhatsgefunden.de"; 

}
so this should be working?

**Reply by jaromirk** [reply]
Yup, looks right!

**Reply by webmasterfcn** [reply]
Can you add script (szenario) for PXE Boot of Windows 10 and 11 (UEFI).

**Reply by jaromirk** [reply]
This? https://github.com/microsoft/MSLab/tree/master/Scenarios/AzSHCI%20and%20MDT

**Comment by webmasterfcn** [comment]
How can i complitly delete the lab. If i run the cleanup there will be still some some files
so do i have to reboot and then can i delete all the files and begin from start?

**Reply by jaromirk** [reply]
What files are left? Cleanup is supposed to revert state to the point before deploy.ps1

**Reply by webmasterfcn** [reply]
Parent disks and other

**Reply by jaromirk** [reply]
that's expected. Cleanup only removes lab from hyper-v, so you can deploy it again (it will only import DC, so deployment process is much faster)

**Comment by webmasterfcn** [comment]
I tried MSLab on windows 11 22H2 with powershell core 7.2.7 and i have problems with the Set context menu option
I installed the msi but the context doesn't work

**Reply by jaromirk** [reply]
@machv  I guess we can remove it from readme, right? @webmasterfcn  context menu is now in native posh install setup as an option.

**Reply by webmasterfcn** [reply]
Can you explain native posh install setup

**Reply by jaromirk** [reply]
Native installer has an option to add context menu https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.2#install-the-msi-package-from-the-command-line. It is also available in GUI when installing PowerShell from MSI.

**Reply by webmasterfcn** [reply]
I have a perfect lab but now my virtual hard disk is to small. How do i extent the size of the virutal disk without losing data?

**Reply by jaromirk** [reply]
In Hyper-V v manager is vhd editor tool. You can expand vhd, and then you just expand partition from inside of VM.

**Comment by webmasterfcn** [comment]
I created the folder d:\folder and copied the mslab scripts in this folder
How can i make that in the virtualmachines the hostname ist the same as the virtual machine name

**Reply by jaromirk** [reply]
In labconfig you have prefix (default is MSLab). If prefix is there, each vm in Hyperv will be MSLab-HostName. If prefix is empty, then name of folder is used. In Your case it would be default-Hostname. It's used like this so you can have different labs in different folders...

**Comment by webmasterfcn** [comment]
l$LabConfig=@{
DomainAdminName='LabAdmin'; AdminPassword='LS1setup!';DCEdition='4';
AdditionalNetworksConfig=@(); VMs=@();
Internet=$true
DomainNetbiosName="CICS";
DomainName="cics.inet";
}
#Management machine
$LabConfig.VMs += @{ VMName = 'Management' ; ParentVHD = 'Win2022_G2.vhdx' }
#Clients
$LabConfig.VMs += @{ VMName = 'Win11' ; ParentVHD = 'Win1122H2_G2.vhdx' }
$LabConfig.VMs += @{ VMName = 'Win10' ; ParentVHD = 'Win1022H2_G2.vhdx' }



what happend here?
`**********************
nStart der Windows PowerShell-Aufzeichnung
Startzeit: 20221213100846
Benutzername: HYPERV-UK\root
RunAs-Benutzer: HYPERV-UK\root
Konfigurationsname: 
Computer: HYPERV-UK (Microsoft Windows NT 10.0.22623.0)
Hostanwendung: C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe -NoProfile -ExecutionPolicy Bypass -File D:\MSLab\3_Deploy.ps1
Prozess-ID: 10480
PSVersion: 5.1.22621.436
PSEdition: Desktop
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.22621.436
BuildVersion: 10.0.22621.436
CLRVersion: 4.0.30319.42000
WSManStackVersion: 3.0
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
**********************
Die Aufzeichnung wurde gestartet. Die Ausgabedatei ist "D:\MSLab\Deploy.log".
Script started at 12/13/2022 10:08:46

MSLab Version dev

Lab telemetry
By providing a telemetry information you will help us to improve MSLab scripts. There are two levels of a telemetry information and we are not collecting any personally identifiable information (PII).
Details about telemetry levels and the content of telemetry messages can be found in documentation https://aka.ms/mslab/telemetry
Available telemetry levels are:
 * None  -- No information will be sent
 * Basic -- Information about lab will be sent (e.g. script execution time, number of VMs, guest OSes)
 * Full  -- Information about lab and the host machine (e.g. type of disk)
Would you be OK with providing an information about your MSLab usage?

Tip: You can also configure telemetry settings explicitly in LabConfig.ps1 file or by setting an environmental variable and suppress this prompt.

No telemetry information will be sent.
List of variables used
	 Prefix used in lab is MSLab-
	 Switchname is MSLab-LabSwitch
	 Workdir is D:\MSLab
	 LabFolder is D:\MSLab\LAB
Checking if OS is Windows 10 1511 (10586)/Server 2016 or newer
	 OS is Windows 10 1511 (10586)/Server 2016 or newer
Checking if Hyper-V is installed
	 Hyper-V is Installed
Checking if Hyper-V Powershell module is installed
	 Hyper-V is Installed
Checking if at least 2GB RAM is available
	 10.219 MB RAM Available
Checking if volume filesystem is NTFS or ReFS
	 Volume filesystem is NTFS
Creating Switch
	 Checking if MSLab-LabSwitch already exists...
	 Creating MSLab-LabSwitch...

Internet connectivity requested
	 Detecting default vSwitch
	 Default switch detected
Checking if lab already exists.
	 Starting IP for AdditionalNetworks is 1
Creating Mountdir
Creating VMs dir
Looking for Tools Parent Disks
	 Tools parent disk not found. Will create one.
Creating Tools.vhdx
Configuring DC
	 Looking for DC to be imported
PS>TerminatingError(Get-VM): "Das Argument für den Parameter "Id" kann nicht überprüft werden. Das Argument ist NULL. Geben Sie einen gültigen Wert für das Argument an, und führen Sie den Befehl erneut aus."
Get-VM : Das Argument für den Parameter "Id" kann nicht überprüft werden. Das Argument ist NULL. Geben Sie einen 
gültigen Wert für das Argument an, und führen Sie den Befehl erneut aus.
In D:\MSLab\3_Deploy.ps1:1230 Zeichen:34
+                 $vm = Get-VM -Id $_.BaseName -ErrorAction SilentlyCon ...
+                                  ~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Get-VM], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Microsoft.HyperV.PowerShell.Commands.GetVM
PS>TerminatingError(Import-VM): "Das Argument für den Parameter "Path" kann nicht überprüft werden. Das Argument ist NULL oder leer. Geben Sie ein Argument an, das nicht NULL oder leer ist, und führen Sie den Befehl erneut aus."
Import-VM : Das Argument für den Parameter "Path" kann nicht überprüft werden. Das Argument ist NULL oder leer. Geben 
Sie ein Argument an, das nicht NULL oder leer ist, und führen Sie den Befehl erneut aus.
In D:\MSLab\3_Deploy.ps1:1237 Zeichen:43
+                     $DC = Import-VM -Path $_.FullName
+                                           ~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Import-VM], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Microsoft.HyperV.PowerShell.Commands.ImportVM
DC was not imported successfully Press any key to continue ...
Press enter to continue ...
**********************
Ende der Windows PowerShell-Aufzeichnung
Endzeit: 20221213100914
**********************`

**Reply by jaromirk** [reply]
Hmm, DC was created successfully? Check lab folder. There should be folder DC with Vhds and vm config.

**Comment by webmasterfcn** [comment]
The Parent Disk are created but then i deploy it got lost. Virtualisierungshost was windows 11 Pro
I think its a problem with windows 11?
Can you try this lab in windows 11
$LabConfig=@{
DomainAdminName='LabAdmin'; AdminPassword='LS1setup!';DCEdition='4';
AdditionalNetworksConfig=@(); VMs=@();
Internet=$true
DomainNetbiosName="JAKA";
DomainName="jaka.inet";
}
#Management machine
$LabConfig.VMs += @{ VMName = 'Management' ; ParentVHD = 'Win2022_G2.vhdx' }
#Clients
$LabConfig.VMs += @{ VMName = 'Win11' ; ParentVHD = 'Win1122H2_G2.vhdx' }
$LabConfig.VMs += @{ VMName = 'Win10' ; ParentVHD = 'Win1022H2_G2.vhdx' }

**Reply by jaromirk** [reply]
can you ping me on Teams? jaromir.kaspar@dell.com

**Comment by webmasterfcn** [comment]
It now works. I did the execute the parent disk in the wrong folder

**Comment by webmasterfcn** [comment]
can you make a Lab like this https://learn.microsoft.com/de-de/azure-stack/hci/deploy/tutorial-private-forest without using cloud services.
No Azure use

**Reply by jaromirk** [reply]
Yup https://geos.to/AzSLabs

**Reply by webmasterfcn** [reply]
can you post the ipconfig /all of the DC

Microsoft Windows [Version 10.0.20348.1366]
(c) Microsoft Corporation. All rights reserved.


Do we have change anything here ? Should we use only ipv4 ?
C:\Windows\system32>ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : DC
   Primary Dns Suffix  . . . . . . . : hvuk.inet
   Node Type . . . . . . . . . . . . : Mixed
   IP Routing Enabled. . . . . . . . : Yes
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : hvuk.inet
                                       mshome.net

Ethernet adapter Ethernet:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Microsoft Hyper-V Network Adapter
   Physical Address. . . . . . . . . : 00-15-5D-1A-8C-2D
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::9215:b2b9:497b:8af1%5(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.0.0.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . :
   DHCPv6 IAID . . . . . . . . . . . : 100668765
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-2B-34-E9-4E-00-15-5D-1A-8C-2D
   DNS Servers . . . . . . . . . . . : ::1
                                       127.0.0.1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Ethernet 2:

   Connection-specific DNS Suffix  . : mshome.net
   Description . . . . . . . . . . . : Microsoft Hyper-V Network Adapter #2
   Physical Address. . . . . . . . . : 00-15-5D-1A-8C-2E
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::db83:bfaf:6b59:4dd8%15(Preferred)
   IPv4 Address. . . . . . . . . . . : 172.26.147.53(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.240.0
   Lease Obtained. . . . . . . . . . : Thursday, December 22, 2022 8:12:21 AM
   Lease Expires . . . . . . . . . . : Friday, December 23, 2022 8:12:23 AM
   Default Gateway . . . . . . . . . : 172.26.144.1
   DHCP Server . . . . . . . . . . . : 172.26.144.1
   DHCPv6 IAID . . . . . . . . . . . : 167777629
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-2B-34-E9-4E-00-15-5D-1A-8C-2D
   DNS Servers . . . . . . . . . . . : 172.26.144.1
   NetBIOS over Tcpip. . . . . . . . : Enabled

C:\Windows\system32>

**Comment by webmasterfcn** [comment]
Hello, i want to test the PRTG in Hyper-V with Server-Core
What is the best method to Download https://www.paessler.com/de/download/prtg-download?download=1^
Should i use Powershell or another method?

**Reply by jaromirk** [reply]
you can just do it like this 

```PowerShell
Start-BitsTransfer -Source https://downloads.paessler.com/prtg/prtg_installer.exe -Destination $env:userprofile\Downloads\prtg_installer.exe
 
```

**Comment by webmasterfcn** [comment]
I want to add a machine which shouldn't join the domain
$LabConfig.VMs += @{ VMName = 'NotInDomain' ; ParentVHD = 'Win2022_G2.vhdx' }
how can i set this and create a local account and extra password

how can i add linux machine also?

**Reply by jaromirk** [reply]
Here is an example for nodjoin https://github.com/DellGEOS/AzureStackHOLs/tree/main/lab-guides/08-DeployAzureStackHCICluster-DeploymentTool#labconfig

**Reply by jaromirk** [reply]
Unattend="Nodjoin"

**Reply by webmasterfcn** [reply]
How can i add local accounts to lab clients?
